source/brailleDisplayDrivers/freedomScientific.py

```diff
@@ -1,292 +1,484 @@
 
-	def _get_numCells(self):
-		return fbGetCellCount(self.fbHandle)
+		self._configureDisplay()
+		self.gestureMap.add("br(freedomScientific):topRouting1",
+			"globalCommands", "GlobalCommands", "braille_scrollBack")
+		self.gestureMap.add("br(freedomScientific):topRouting%d" % self.numCells,
+			"globalCommands", "GlobalCommands", "braille_scrollForward")
 
-	def display(self,cells):
-		cells="".join([chr(x) for x in cells])
-		fbWrite(self.fbHandle,0,len(cells),cells)
+	def terminate(self):
+		try:
+			super(BrailleDisplayDriver, self).terminate()
+		finally:
+			# Make sure the device gets closed.
+			# If it doesn't, we may not be able to re-open it later.
+			self._dev.close()
+
+	def _sendPacket(self, packetType, arg1=FS_BYTE_NULL, arg2=FS_BYTE_NULL, arg3=FS_BYTE_NULL, data=FS_DATA_EMPTY):
+		"""Send a packet to the display
+
+		@param packetType: Type of packet (first byte), use one of the FS_PKT constants
+		@type packetType: str
+		@param arg1: First argument (second byte of packet)
+		@type arg1: str
+		@param arg2: Second argument (third byte of packet)
+		@type arg2: str
+		@param arg3: Third argument (fourth byte of packet)
+		@type arg3: str
+		@param data: Data to send if this is an extended packet, required checksum will be added automatically
+		@type data: str
+		"""
+		def handleArg(arg):
+			if type(arg) == int:
+				return int2byte(arg)
+			return arg
+		arg1 = handleArg(arg1)
+		arg2 = handleArg(arg2)
+		arg3 = handleArg(arg3)
+		packet = [packetType, arg1, arg2, arg3, data]
+		if data:
+			packet.append(int2byte(self._calculateChecksum("".join(packet))))
+		self._dev.write("".join(packet))
+
+	def _onReceive(self, data):
+		if self.isUsb:
+			data = BytesIO(data)
+			packetType = data.read(1)
+		else:
+			packetType = data
+			data = self._dev
+
+		arg1 = data.read(1)
+		arg2 = data.read(1)
+		arg3 = data.read(1)
+		log.debug("Got packet of type %r with args: %r %r %r", packetType, arg1, arg2, arg3)
+		# Info and extended key responses are the only packets with payload and checksum
+		if packetType in (FS_PKT_INFO, FS_PKT_EXT_KEY):
+			length = ord(arg1)
+			payload = data.read(length)
+			checksum = ord(data.read(1))
+			calculatedChecksum = self._calculateChecksum(packetType + arg1 + arg2 + arg3 + payload)
+			assert calculatedChecksum == checksum, "Checksum mismatch, expected %s but got %s" % (checksum, payload[-1])
+		else:
+			payload = FS_DATA_EMPTY
+
+		self._handlePacket(packetType, arg1, arg2, arg3, payload)
+
+	def _handlePacket(self, packetType, arg1, arg2, arg3, payload):
+		"Handle a packet from the device"
+		if packetType == FS_PKT_ACK:
+			self._handleAck()
+		elif packetType == FS_PKT_NAK:
+			log.debugWarning("NAK received!")
+			self._handleAck()
+		elif packetType == FS_PKT_INFO:
+			self._manufacturer = payload[:24].replace(FS_DATA_EMPTY, "")
+			self._model = payload[24:40].replace(FS_DATA_EMPTY, "")
+			self._firmwareVersion = payload[40:48].replace(FS_DATA_EMPTY, "")

```

Comment:
These should be `FS_BYTE_NULL`, not `FS_DATA_EMPTY`. This is breaking the driver at least for me.